gax/src/main/java/com/google/api/gax/tracing/OpencensusTracer.java

```diff
@@ -0,0 +1,372 @@
+
+    span.putAttributes(attributes);
+    span.end();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void operationCancelled() {
+    Map<String, AttributeValue> attributes = baseOperationAttributes();
+    span.putAttributes(attributes);
+    span.end(
+        EndSpanOptions.builder()
+            .setStatus(Status.CANCELLED.withDescription("Cancelled by caller"))
+            .build());
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void operationFailed(Throwable error) {
+    Map<String, AttributeValue> attributes = baseOperationAttributes();
+
+    span.putAttributes(attributes);
+    span.end(EndSpanOptions.builder().setStatus(convertErrorToStatus(error)).build());
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void connectionSelected(int id) {
+    span.addAnnotation(
+        "Connection selected", ImmutableMap.of("id", AttributeValue.longAttributeValue(id)));
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptStarted(int attemptNumber) {
+    currentAttemptId = attemptNumber;
+    attemptSentMessages = 0;
+    attemptReceivedMessages = 0;
+
+    HashMap<String, AttributeValue> attributes = Maps.newHashMap();
+    populateAttemptNumber(attributes);
+
+    span.addAnnotation("Attempt started", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptSucceeded() {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+
+    span.addAnnotation("Attempt succeeded", attributes);
+  }
+
+  @Override
+  public void attemptCancelled() {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+
+    span.addAnnotation("Attempt cancelled", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptFailed(Throwable error, Duration delay) {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    attributes.put("delay ms", AttributeValue.longAttributeValue(delay.toMillis()));
+    populateError(attributes, error);
+
+    String msg = error != null ? "Attempt failed" : "Operation incomplete";
+    span.addAnnotation(msg + ", scheduling next attempt", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptFailedRetriesExhausted(Throwable error) {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    populateError(attributes, error);
+
+    span.addAnnotation("Attempts exhausted", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptPermanentFailure(Throwable error) {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    populateError(attributes, error);
+
+    span.addAnnotation("Attempt failed, error not retryable", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void responseReceived() {
+    attemptReceivedMessages++;
+    totalReceivedMessages++;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void requestSent() {
+    attemptSentMessages++;
```

Comment:
I'm not sure what  you mean by adding noise with synchronization here. It is really unnerving to keep a class which will be initizlized/modified/consumed by various threads unsynchronzied (I did read the IBM's article about thread safety categories). 

Well, we always can add some "thread friendliness" by doing something like the following:
1) Make fields (like `attemptSentMessages`) at at least `volatile` (to ensure thread visibility).
2) Or Make them `Atomic*` (visibility + atomic increments of each field)
3) Or Synchronize the whole class (could be done in an old-fashioned way of adding `synchronize` keyword to each method signature, which can greatly reduce amount of boilerplate code, if that is what you meant by noise) (to ensure atomicity of the whole thing).

In any case, this is a reporting class, which does not affect decision-making process and even if updates to it stay non-atomic and/or are not properly syncronized by the outer code the worst thing which can happen is slightly wrong report numbers. Taking that into account can you at least make the non-final fields of method at least volatile (better Anomic) to make me sleep better? 

As for performance effect, we already do locking, so it won't change anything drastically. According to [Jeff Dean's numbers](https://gist.github.com/jboner/2841832) Mutex lock/unlock is about 25ns (and those are quite old, the google's internal version of the table says around 17ns now). In any case it is quite cheap, especially taking into account how much an actual network call takes.

Reply:
In this scenario, there are only 2 threads that are relevant: the caller thread & the grpc thread. The grpc thread is responsible for closing spans and generating annotations.

So the counters that are only modified by the grpc thread dont need any special handling. The other counters (for sending msgs & tracking the attempt number) due need some consideration. I marked them as volatile. Thanks for catching that.